PDEP-13: Deprecate the apply method on Series and DataFrame and make the agg and transform methods operate on series data (#54747)
pandas-dev/pandas: web/pandas/pdeps/0013-standardize-apply.md

It is proposed that `apply`, `transform` and `agg` in the future will work as follows:

1. the `agg` & `transform` methods of `Series`, `DataFrame` & `groupby` will always operate series-wise and never element-wise
2. `Series.apply` & `DataFrame.apply` will be deprecated.
3. `groupby.apply` will not be deprecated (because it behaves differently than `Series.apply` & `DataFrame.apply`)

The above changes means that the future behavior, when users want to apply arbitrary callables in pandas, can be described as follows:

1. When users want to operate on single elements in a `Series` or `DataFrame`, they should use `Series.map` and `DataFrame.map` respectively.
2. When users want to aggregate a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.agg`, `DataFrame.agg` and `groupby.agg` respectively.

> 3. When users want to transform a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.transform`, `DataFrame.transform` and `groupby.transform` respectively.

Reviewer: Could you add a 4th "When users want to neither aggregate or transform a ..., they should ..."

> 3. When users want to transform a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.transform`, `DataFrame.transform` and `groupby.transform` respectively.

Reviewer 2: I think you need to have the case when users want to have an operation on a row of a `DataFrame`, they should use `DataFrame.agg(func, axis=1)` .  Today a common use case is using `DataFrame.apply(func, axis=1)`.  Since `func` can do anything, it isn't entirely clear that `agg` with `axis=1` will supply each row of the series to the function, meaning that the corresponding function doesn't need to "aggregate".

Author: `DataFrame.apply(func, axis=1)` is just `DataFrame.T.apply(func).T`, so any issue with `axis=1` will also be present with `axis=0`.  So I don't see a special issue with `axis=1`, except maybe some idioms that are more common with `axis=1` (which may be a valid concern, though).

Can you give some example code snippets showing some possible problems?

Reviewer 3: @Dr-Irv I've added an example here https://github.com/pandas-dev/pandas/pull/54747/files#r1339884642

Inclined to agree that `.agg` feels really unnatural here

> 3. When users want to transform a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.transform`, `DataFrame.transform` and `groupby.transform` respectively.

Reviewer 2: I'm not convinced of the utility of `Series.transform` or `DataFrame.transform`.  Can those functions do anything that `.map()` and `.agg()` can't do?

If not, maybe we should deprecate `transform` as well?

Author: They do stuff like `df.transform(lambda x: x - x.mean())`, which is not good to do using `map` or `agg`.  Not useful all that often, but is useful in some important cases though.

Reviewer 2: > They do stuff like `df.transform(lambda x: x - x.mean())`, which is not good to do using `map` or `agg`. Not useful all that often, but is useful in some important cases though.

Good example.

> The use of `Series.apply` &  `DataFrame.apply` will after the proposed change in almost all cases be replaced by `map`, `agg` or `transform`. In the very few cases where `Series.apply` &  `DataFrame.apply` cannot be substituted by `map`, `agg` or `transform`, it is proposed that it will be accepted that users will have to find alternative ways to apply the functions, i.e. typically apply the functions manually and possibly concatenating the results.

Reviewer 2: It may be useful to have some examples of those "very few cases"

## Motivation

The current behavior of `apply`, `agg` & `transform` is very complex and therefore difficult to understand for non-expert users. The difficulty is especially that the methods sometimes apply callables on elements of series/dataframes, sometimes on Series or columns/rows of Dataframes and sometimes try element-wise operation and if that fails, falls back to series-wise operations.

Below is an overview of the current behavior in table form when giving callables to `agg`, `transform` & `apply`. As an example on how to read the tables, when a non-ufunc callable is given to `Series.agg`, `Series.agg` will first try to apply the callable to each element in the series, and if that fails, will fall back to call the series using the callable.

> (The description may not be 100 % accurate because of various special cases in the current implementation, but will give a good understanding of the current behavior).

Reviewer 2: should you add the string representations of functions to the tables?

Author: The problems that I try to solve only concern callables and I try to keep the PDEP brief. I can coment somewhere that the proposal does not affect supplying strings to the methods.

Reviewer 2: But even the behavior with strings is confusing.  Here's some examples to illustrate:
```python
>>> s = pd.Series([1,2,3])
>>> s.apply("sum")
6
>>> s.apply(np.sum)
0    1
1    2
2    3
dtype: int32
>>> s.apply("mean")
2.0
>>> s.apply(np.mean)
0    1.0
1    2.0
2    3.0
dtype: float64
>>> s.apply(abs)
0    1
1    2
2    3
dtype: int64
>>> s.apply(np.abs)
0    1
1    2
2    3
dtype: int64
>>> s.apply("abs")
0    1
1    2
2    3
dtype: int64
```

         A
  <lambda>
0        0
1        1
2        2
>>> df.A.agg(lambda x: np.sum(x))  # not ok
0    0
1    1
2    2
Name: A, dtype: int64
```

It can also have great effect on performance, even when the result is correct. For example:

```python
>>> df = pd.DataFrame({"A": range(1_000_000)})
>>> %tiemit df.transform(lambda x: x + 1)  # fast
1.43 ms ± 3.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
 >>> %timeit df.transform([lambda x: x + 1])  # slow
163 ms ± 754 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
 >>> %timeit df.A.transform(lambda x: x + 1)  # slow
162 ms ± 980 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

The reason for the great performance difference is that `df.transform(func)` operates on series data, which is fast, while `df.transform(func_list)` will attempt elementwise operation first, and if that works (which is does here), will be much slower than series operations.

In addition to the above effects of the current implementation of `agg`/`transform` & `apply`, see [#52140](https://github.com/pandas-dev/pandas/issues/52140) for more examples of the unexpected effects of how `apply` is implemented.

> It can also be noted that `Series.apply` & `DataFrame.apply` could almost always be replaced with calls to `agg`, `transform` & `map`, if `agg` & `transform` were to always operate on series data. For some examples, see the table below for alternatives using `apply(func)`:

Reviewer 2: ```suggestion
It can also be noted that `Series.apply` & `DataFrame.apply` could almost always be replaced with calls to `agg`, `transform` or `map`, if `agg` and `transform` were to always operate on series data. For some examples, see the table below for alternatives using `apply(func)`:
```

| func                    | Series     | DataFrame   |
|:--------------------|:-----------|:------------|
| lambda x: str(x)    | .map       | .map        |
| lambda x: x + 1     | .transform | .transform  |
| [lambda x: x.sum()] | .agg       | .agg        |

> So, for example, `ser.apply(lambda x: str(x))` can be replaced with `ser.map(lambda x: str(x))` while `df.apply([lambda x: x.sum()])` can be replaced with `df.agg([lambda x: x.sum()])`.

Reviewer 2: Would be really worthwhile to include some `axis=1` examples for `DataFrame.apply()`

Overall, because of their flexibility, `Series.apply` & `DataFrame.apply` are considered unnecessarily complex, and it would be clearer for users to use `.map`, `.agg` or `.transform`, as appropriate in the given situation.

## Proposal

With the above in mind, it is proposed that in the future:

> It is proposed that `apply`, `transform` and `agg` in the future will work as follows:

Reviewer 2: ```suggestion
It is proposed that `apply`, `transform` and `agg` will work as follows:
```

(You have "in the future" in 2 consecutive sentences)

1. the `agg` & `transform` methods of `Series`, `DataFrame` & `groupby` will always operate series-wise and never element-wise
2. `Series.apply` & `DataFrame.apply` will be deprecated.
3. `groupby.apply` will not be deprecated (because it behaves differently than `Series.apply` & `DataFrame.apply`)

The above changes means that the future behavior, when users want to apply arbitrary callables in pandas, can be described as follows:

1. When users want to operate on single elements in a `Series` or `DataFrame`, they should use `Series.map` and `DataFrame.map` respectively.
2. When users want to aggregate a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.agg`, `DataFrame.agg` and `groupby.agg` respectively.
3. When users want to transform a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.transform`, `DataFrame.transform` and `groupby.transform` respectively.

> The use of `Series.apply` &  `DataFrame.apply` will after the proposed change in almost all cases be replaced by `map`, `agg` or `transform`. In the very few cases where `Series.apply` &  `DataFrame.apply` cannot be substituted by `map`, `agg` or `transform`, it is proposed that it will be accepted that users will have to find alternative ways to apply the functions, i.e. typically apply the functions manually and possibly concatenating the results.

Reviewer 2: this text seems to be repeated from the abstract.  So shorten the abstract - leave the details here

> It can be noted that `groupby.agg`, `groupby.transform` & `groupby.apply` are not proposed changed in this PDEP, because `groupby.agg`, `groupby.transform` already behave as desired and `groupby.apply` behaves differently than `Series.apply` & `DataFrame.apply`. Likewise, the behavior when given ufuncs will remain unchanged, because the behavior is already as intended in all cases.

Reviewer 2: ```suggestion
It can be noted that the behavior of `groupby.agg`, `groupby.transform` and `groupby.apply` are not proposed to be changed in this PDEP, because `groupby.agg`, `groupby.transform` already behave as desired and `groupby.apply` behaves differently than `Series.apply` & `DataFrame.apply`. Likewise, the behavior when given ufuncs will remain unchanged, because the behavior is already as intended in all cases.
```

## Deprecation process

To change the current behavior, it will have to be deprecated. This will be done by in v2.2:

1. Deprecate `Series.apply` & `DataFrame.apply`.

> 2. Add a `series_ops_only` with type `bool | lib.NoDefault` parameter to `agg` & `transform` methods of `Series` & `DataFrame`. When `series_ops_only` is set to False, `agg` & `transform` will behave as they do currently. When set to True, `agg` & `transform` will never operate on elements, but always on Series. When set to `no_default`, `agg` & `transform` will behave as `series_ops_only=False`, but will emit a FutureWarning the current behavior will be reoved in the future.

Reviewer 2: ```suggestion
2. Add a `series_ops_only` argument with type `bool | lib.NoDefault` parameter to `agg` and `transform` methods of `Series` and `DataFrame` with a default value of `lib.NoDefault`.  When `series_ops_only` is set to `False`, `agg` and  `transform` will behave as they do currently. When set to `True`, `agg` and `transform` will never operate on elements, but always on Series. When set to `no_default`, `agg` and  `transform` will behave as `series_ops_only=False`, but will emit a `FutureWarning` the current behavior will be removed in the future.
```

- Status: Under discussion
- Discussion: [#52140](https://github.com/pandas-dev/pandas/issues/52509)
- Author: [Terji Petersen](https://github.com/topper-123)
- Revision: 2

## Abstract

The `apply`, `transform` and `agg` methods have very complex behavior when given callables because they in some cases operate on elements in series, in some cases on series and sometimes try one first, and it that fails, falls back to try the other. There is not a logical system how these behaviors are arranged and it can therefore be difficult for users to understand these methods.

It is proposed that `apply`, `transform` and `agg` in the future will work as follows:

1. the `agg` and `transform` methods of `Series`, `DataFrame` and `groupby` will always operate series-wise and never element-wise
2. `Series.apply` and `DataFrame.apply` will be deprecated.
3. The current behavior when supplying string to the methods will not be changed.
4. `groupby.apply` will not be deprecated (because it behaves differently than `Series.apply` and `DataFrame.apply`)

The above changes means that the future behavior, when users want to apply arbitrary callables in pandas, can be described as follows:

> 1. When users want to operate on single elements in a `Series` or `DataFrame`, they should use `Series.map` and `DataFrame.map` respectively.

Reviewer 3: Are we sure this is accurate? For the use cases I've seen of `DataFrame.apply`, I think `DataFrame.agg` would probably be more likely as a replacement

Currently, say I do
```python
In [43]: df = pd.DataFrame({'a': ['quetzal', 'quetzal', 'baboon'], 'b': ['panda', 'chinchilla', 'elk']})

In [44]: df
Out[44]:
         a           b
0  quetzal       panda
1  quetzal  chinchilla
2   baboon         elk

In [45]: def func(row):
    ...:     return f"{row['a']} or {row['b']}?"
    ...:

In [46]: df.apply(func, axis=1)
Out[46]:
0         quetzal or panda?
1    quetzal or chinchilla?
2            baboon or elk?
dtype: object
```

Then if I use `DataFrame.map`, it'll operate over all elements of the dataframe 1-by-1. Whereas if I use `.agg`, then I'll preserve what I'm currently doing:
```python
In [47]: df.agg(func, axis=1)
Out[47]:
0         quetzal or panda?
1    quetzal or chinchilla?
2            baboon or elk?
dtype: object
```

Reviewer 4: Isn't this a transform? But users don't get to control what gets passed to the UDF they supply - is it columm by column, row by row, or the entire frame. Perhaps they should.

Reviewer 3: transform would fail here:
```python
In [53]: df.transform(func, axis=1)
---------------------------------------------------------------------------

ValueError: Function did not transform
```

And this underscores just how confusing this is

> is it columm by column, row by row, or the entire frame. Perhaps they should.

yes, exactly, it could be:
`axis=None`: element by element (current behaviour)
`axis=1`: row by row
`axis=0`: element by element

Then, `df.apply(lambda row: ..., axis=1)` can just become `df.map(lambda row: ..., axis=1)`

This is a very common way of using `apply`, e.g.
```python
titanic_df['Person'] = titanic_df[['Age','Sex']].apply(get_person,axis=1)
```
from https://www.kaggle.com/code/omarelgabry/a-journey-through-titanic

Author: @MarcoGorelli , your example is an aggregation though and `func` does operate on rows, not elements, so everything works as intended in your `agg` example.

However, I guess `agg` linguistically(?) is best understood as a numerical reduction, not a text combination operation, so using `agg` may be a bit unintuitive. My original proposal would have kept `apply` for situations like this, because the problem I wanted to solve was   `apply` giving subtly different results for `Series.apply` and `DataFrame.apply` (including for `axis=1`).

Your example could be an argument for keeping `apply`, but drop the element-by-element behavior of `Series.apply`, i.e. revert to the original proposal of this PDEP.

I'm -1 on making `map` operate on rows/columns. It's a strength of the method that it  always gives same-shaped results and it would be confusing for the result changing shape based on if an axis parameter is given or not.

Reviewer 3: true, it's a horizontal aggregation: it combines multiple columns into 1

I'm just thinking about the upgrade process

I've only ever seen people use `DataFrame.apply` to create a single column, and if the upgrade process is going to be:
- Series.apply(func) => Series.map(func)
- DataFrame.apply(func, axis=1) => DataFrame.agg(func, axis=1)

then that could be confusing, and certainly harder than just grepping for `.apply`

Reviewer 3: wait let's not add yet another method to the API 😅

Reviewer 5: (just another voice here to support the notion that requiring `agg` for this use case feels very unnatural)

2. When users want to aggregate a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.agg`, `DataFrame.agg` and `groupby.agg` respectively.
3. When users want to transform a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.transform`, `DataFrame.transform` and `groupby.transform` respectively.
4. Functions that are not applicable for  `map`, `agg` nor `transform` are considered relatively rare and in the future users should call these functions directly rather than use the `apply` method.

The use of `Series.apply` and  `DataFrame.apply` will after the proposed change in almost all cases be replaced by `map`, `agg` or `transform`. In the very few cases where `Series.apply` and  `DataFrame.apply` cannot be substituted by `map`, `agg` or `transform`, it is proposed that it will be accepted that users will have to find alternative ways to apply the functions, i.e. typically apply the functions manually and possibly concatenating the results.

## Motivation

> The current behavior of `apply`, `agg` and `transform` is very complex and therefore difficult to understand for non-expert users. The difficulty is especially that the methods sometimes apply callables on elements of series/dataframes, sometimes on Series or columns/rows of Dataframes and sometimes try element-wise operation and if that fails, falls back to series-wise operations.

Reviewer 6: I find the motivation to be pretty strong from a developer perspective but I think it is lacking from an end user perspective. To play devil's advocate...why should a non-expert user care or try to understand this distinction?

While there are definitely warts here part of me thinks we've had them for X number of years and gotten by I think without too much end user complaint about it. I think this would be the largest deprecation we've done on the project as long as I've been involved, so I'm a little wary of causing churn instead of just trying to "softly" guide users towards the more explicit map / agg / transform options

|:-----------------------------------|:---------------------------------|:---------------------------------|:----------|
| ufunc or list/dict of ufuncs       | series                           | series                           | series    |
| other callables (non ufunc)        | Try elements, fallback to series | series                           | series    |
| list/dict of callables (non-ufunc) | Try elements, fallback to series | Try elements, fallback to series | series    |

### apply

|                                    | Series   | DataFrame   | groupby   |
|:-----------------------------------|:---------|:------------|:----------|
| ufunc or list/dict of ufuncs       | series   | series      | series    |
| other callables (non ufunc)        | elements | series      | series    |
| list/dict of callables (non-ufunc) | Try elements, fallback to series | series    | series    |

The 3 tables show that:

1. when given numpy ufuncs, callables given to `agg`/`transform`/`apply` operate on series data
2. when used on groupby objects, callables given to `agg`/`transform`/`apply` operate on series data

> 3. else, in some case it will try element-wise operation and fall back to series-wise operations if that fails, in some case will operate on series data and in some cases on element data.

Reviewer 5: Sidenote: to be honest, this "in some cases on element data" is IMO by far the _main_ use case of `Series.apply`, and in this case (being passed a single function), it always operates element by element, consistently and unambiguously.

1. When users want to operate on single elements in a `Series` or `DataFrame`, they should use `Series.map` and `DataFrame.map` respectively.
2. When users want to aggregate a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.agg`, `DataFrame.agg` and `groupby.agg` respectively.
3. When users want to transform a `Series`, columns/rows of a `DataFrame` or groups in `groupby` objects, they should use `Series.transform`, `DataFrame.transform` and `groupby.transform` respectively.
4. Functions that are not applicable for  `map`, `agg` nor `transform` are considered relatively rare and in the future users should call these functions directly rather than use the `apply` method.

The use of `Series.apply` and  `DataFrame.apply` will after the proposed change in almost all cases be replaced by `map`, `agg` or `transform`. In the very few cases where `Series.apply` and  `DataFrame.apply` cannot be substituted by `map`, `agg` or `transform`, it is proposed that it will be accepted that users will have to find alternative ways to apply the functions, i.e. typically apply the functions manually and possibly concatenating the results.

> It can be noted that the behavior of `groupby.agg`, `groupby.transform` and `groupby.apply` are not proposed changed in this PDEP, because `groupby.agg`, `groupby.transform` already behave as desired and `groupby.apply` behaves differently than `Series.apply` and `DataFrame.apply`. Likewise, the behavior when given ufuncs (e.g. `np.sqrt`) and string input (e.g. `"sqrt"`) will remain unchanged, because the behavior is already as intended in all cases.

Reviewer 2: ```suggestion
It can be noted that the behavior of `groupby.agg`, `groupby.transform` and `groupby.apply` are not proposed to be changed in this PDEP, because `groupby.agg`, `groupby.transform` already behave as desired and `groupby.apply` behaves differently than `Series.apply` and `DataFrame.apply`. Likewise, the behavior when given ufuncs (e.g. `np.sqrt`) and string input (e.g. `"sqrt"`) will remain unchanged, because the behavior is already as intended in all cases.
```

# PDEP-13: Make the Series.apply method operate Series-wise

- Created: 24 August 2023
- Status: Under discussion
- Discussion: [#52140](https://github.com/pandas-dev/pandas/issues/52509)
- Author: [Terji Petersen](https://github.com/topper-123)
- Revision: 1

## Abstract

Currently, giving a input to `Series.apply` is treated differently depending on the type of the input:

> * if the input is a numpy `ufunc`, `series.apply(func)` is equivalent to `func(series)`, i.e. similar to `series.pipe(func)`.

Author: Strings are a bit different than that, and `series.apply(string)` is more equivalent to `getattr(series, string)()`. I've added a line about strings.

* if the input is a callable, but not a numpy `ufunc`, `series.apply(func)` is similar to `Series([func(val) for val in series], index=series.index)`, i.e. similar to `series.map(func)`

> * if the input is a list-like or dict-like, `series.apply(func)` is equivalent to `series.agg(func)` (which is subtly different than `series.apply`)

Author: I've made it more precise, my intention was not to exclude strings. e.g. `s.apply(["sqrt", np.sqrt])` is also possible.

> * if the input is a list-like or dict-like, `series.apply(func)` is equivalent to `series.agg(func)` (which is subtly different than `series.apply`)

Author: Also, the described behavior is the actual behavior, so e.g. a list of ints will be passed through and will end up calling each int and fail. So the current text is accurate.

In contrast, `DataFrame.apply` has a consistent behavior:

* if the input is a callable, `df.apply(func)` always calls each columns in the DataFrame, so is similar to `func(col) for _, col in
df.items()` + wrapping functionality

> * if the input is a list-like or dict-like, `df.apply` call each item in the list/dict and wraps the result as needed. So for example if the input is a list, `df.apply(func_list)` is equivalent to `[df.apply(func) for func in func_list]` + wrapping functionality

Author: I've added something about this.

This PDEP proposes that:

- The current complex current behavior of `Series.apply` will be deprecated in Pandas 2.2.

> - Single callables given to the `.apply` methods of `Series` will in Pandas 3.0 always be called on the whole `Series`, so `series.apply(func)` will become similar to `func(series)`,

Reviewer 7: 4 months is way too short of a deprecation period for a change that will most likely affect the majority of our users

Author: I'm a bit surprised about this. I would expect users who upgrade major versions to run their code through the latest minor version before upgrading and if they get a warning, to fix them before upgrading. Same for libraries, that they would check and fix warnings before upgrading support for major versions.

Only implementing this in v4.0 seems way too slow IMO, if we decide this is a good idea in itself.

Reviewer 3: @phofl are you sure about 

> There is no easy way to opt into the new behaviour.

? I thought the way to do that was by passing `by_row=True`?

> - Lists or dicts of callables given to the `Series.apply` will in Pandas 3.0 always call `Series.apply` on each element of the list/dict

Author: That won't expand list or dicts to Dataframes, e.g. `ser.apply(["sqrt", np.sqrt])` is a dataframe.

I've made a new formulation to clarify this.

`Series.apply` & `Series.map` currently often behave very similar, but differently enough for it to be confusing when it's a good idea to use one over the other and especially when `Series.apply` is a bad idea to use.

Also, calling `Series.apply` currently gives a different result than the per-column result from calling `DataFrame.apply`, which can be confusing for users who expect `Series.apply` to be the `Series` version of `DataFrame.apply`, similar to how `Series.agg` is the `Series` version of `DataFrame.agg`. For example, currently some functions may work fine with `DataFrame.apply`, but may fail, be very slow when given to `Series.apply` or give a different result than the per-column result from `DataFrame.apply`.

### Similarities and differences between `Series.apply` and `Series.map`

The similarity between the methods is especially that they both fall back to use `Series._map_values` and there use `algorithms.map_array` or `ExtensionArray.map` as relevant.

The differences are many, but each one is relative minor:

1. `Series.map` has a `na_action` parameter, which `Series.apply` doesn't

> 2. `Series.apply` can take advantage of numpy ufuncs, which `Series.map` can't

Reviewer 3: I read this as "`apply` can efficiently make use of numpy ufuncs, whereas `map` has to run them element-by-element", which seems correct

3. `Series.apply` can take `args` and `**kwargs`, which `Series.map` can't
4. `Series.apply` is more general and can take a string, e.g. `"sum"`, or lists or dicts of inputs which `Series.map` can't.

> 5. when given a numpy ufunc, the ufunc will be called on the whole Series, when given to `Series.apply` and on each element of the series, if given to `Series.map`.

Reviewer 3: how so?

```python
>>> %timeit large_ser.pipe(lambda x: x + 1)
44 µs ± 363 ns per loop
```

(The reason for the above performance differences is that apply gets called on each single element, while `pipe` calls `x.__add__(1)`, which operates on the whole array).

Note also that `.pipe` operates on the `Series` while `apply`currently operates on each element in the data, so there is some differences that may have some consequence in some cases.

This PDEP proposes that callables will be applies to the whole `Series`, so we in the future `Series.apply` will be as fast as `Series.pipe`.

### 4. ufuncs in `Series.apply` vs. normal functions

Performance-wise, ufuncs are fine in `Series.apply`, but non-ufunc functions are not:

```python
>>> %timeit large_ser.apply(np.sqrt)
71.6 µs ± 1.17 µs per loop
>>> %timeit large_ser.apply(lambda x:np.sqrt(x))
63.6 ms ± 1.03 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

It is difficult to understand why ufuncs are fast in `apply`, while other callables are slow in `apply` (answer: it's because ufuncs operate on the whole Series, while other callables operate elementwise).

> This PDEP proposes that callables will be applies to the whole `Series`, so we in the future non-ufunc functions in `Series.apply` will be as fast as ufuncs.

Reviewer 3: general comment, any chance we leave these kinds of review comments until the end, when the substance of the pdep has already been discussed and agreed-upon? thanks 🙏

### 5. callables in `Series.apply` is slow, callables in `DataFrame.apply` is fast

Above it was shown that using (non-ufunc) callables in `Series.apply` is bad performance-wise. OTOH using them in `DataFrame.apply` is fine:

```python
>>> %timeit large_ser.apply(lambda x: x + 1)
24.3 ms ± 24 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
>>> %timeit pd.DataFrame(large_ser).apply(lambda x: x + 1)
160 µs +- 1.17 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Having callables being fast to use in the `DataFrame.apply` method, but slow in `Series.apply` is confusing for users.

This PDEP proposes that callables will be applies to the whole `Series`, so we in the future `Series.apply` will be as fast as `DataFrame.apply` already is.

### 6. callables in `Series.apply` may fail, while callables in `DataFrame.apply` do not and vica versa

```python
>>> ser.apply(lambda x: x.sum())
AttributeError: 'int' object has no attribute 'sum'
>>> pd.DataFrame(ser).apply(lambda x: x.sum())
0    6
dtype: int64

> ```

Reviewer 2: You should show the reverse example:
```python
In [27]: small_ser.apply(lambda x: x+3 if x<2 else x*10)
Out[27]:
0     4
1    20
2    30
dtype: int64

In [28]: pd.DataFrame(small_ser).apply(lambda x: x+3 if x<2 else x*10)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-28-7950464b3dd1> in <module>
----> 1 pd.DataFrame(small_ser).apply(lambda x: x+3 if x<2 else x*10)

C:\Anaconda3\lib\site-packages\pandas\core\frame.py in apply(self, func, axis, raw, result_type, args, **kwargs)
   8734             kwargs=kwargs,
   8735         )
-> 8736         return op.apply()
   8737
   8738     def applymap(

C:\Anaconda3\lib\site-packages\pandas\core\apply.py in apply(self)
    686             return self.apply_raw()
    687
--> 688         return self.apply_standard()
    689
    690     def agg(self):

C:\Anaconda3\lib\site-packages\pandas\core\apply.py in apply_standard(self)
    810
    811     def apply_standard(self):
--> 812         results, res_index = self.apply_series_generator()
    813
    814         # wrap results

C:\Anaconda3\lib\site-packages\pandas\core\apply.py in apply_series_generator(self)
    826             for i, v in enumerate(series_gen):
    827                 # ignore SettingWithCopy here in case the user mutates
--> 828                 results[i] = self.f(v)
    829                 if isinstance(results[i], ABCSeries):
    830                     # If we have a view on v, we need to make a copy because

<ipython-input-28-7950464b3dd1> in <lambda>(x)
----> 1 pd.DataFrame(small_ser).apply(lambda x: x+3 if x<2 else x*10)

C:\Anaconda3\lib\site-packages\pandas\core\generic.py in __nonzero__(self)
   1536     def __nonzero__(self):
   1537         raise ValueError(
-> 1538             f"The truth value of a {type(self).__name__} is ambiguous. "
   1539             "Use a.empty, a.bool(), a.item(), a.any() or a.all()."
   1540         )

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```

> Having callables fail when used in `Series.apply`, but work in `DataFrame.Apply` or vica versa is confusing for users.

Reviewer 2: ```suggestion
Having callables fail when used in `Series.apply`, but work in `DataFrame.Apply` or vice versa is confusing for users.
```

> This PDEP proposes that callables will be applies to the whole `Series`, so callables given to `Series.apply` will work the same as when given to `DataFrame.apply`, so in the future we will have that:

Reviewer 2: ```suggestion
This PDEP proposes that callables will be applied to the whole `Series`, so callables given to `Series.apply` will work the same as when given to `DataFrame.apply`, so in the future we will have that:
```

```python
>>> ser.apply(lambda x: x.sum())
6
>>> pd.DataFrame(ser).apply(lambda x: x.sum())
0    6
dtype: int64
```

### 7.  `Series.apply` vs. `Series.agg`

The doc string for `Series.agg` says about the method's `func` parameter: "If a function, must ... work when passed ... to Series.apply". But compare these:

```python
>>> small_ser.agg(np.sum)
6
>>> small_ser.apply(np.sum)
0    1
1    2
2    3
dtype: int64
```

> Users would expect these two to give the same result.

Reviewer 2: Ummm, not so sure about that.  What about the case of `np.abs()`  or just `abs` ?  In that case `Series.agg(abs)` gets applied element-wise!

Author: Those will give the same results, so the text is accurate?

I agree we cannot be guaranteed that `ser.agg(func)` will actually aggregate, but that is IMO a different discussion and maybe part of the discussion that has been opened about changing the API for this class of methods.

> This PDEP proposes that callables will be applies to the whole `Series`, so in the future we will have:

Reviewer 2: ```suggestion
This PDEP proposes that callables will be applied to the whole `Series`, so in the future we will have:
```

```python
>>> small_ser.agg(np.sum)
6
>>> small_ser.apply(np.sum)
6
```

### 8. dictlikes vs. listlikes in `Series.apply`

Giving a *list* of transforming arguments to `Series.apply` returns a `DataFrame`:

```python
>>> small_ser.apply(["sqrt", np.abs])
       sqrt  absolute
0  1.000000         1
1  1.414214         2
2  1.732051         3
```

But giving a *dict* of transforming arguments returns a `Series` with a `MultiIndex`:

```python
>>> small_ser.apply({"sqrt" :"sqrt", "abs" : np.abs})
sqrt  0    1.000000
      1    1.414214
      2    1.732051
abs   0    1.000000
      1    2.000000
      2    3.000000
dtype: float64
```

> These two should give same-shaped output for consistency. Using `Series.transform` instead of `Series.apply`, it returns a `DataFrame` in both cases and I think the dictlike example above should return a `DataFrame` similar to the listlike example.

Reviewer 2: I agree with you here that this inconsistency is hard to understand.

3. The changes made to `Series.apply`will propagate to `Series.agg` and `Series.transform` as needed.

The difference between `Series.apply()` & `Series.map()` will then be that:

* `Series.apply()` makes the passed-in callable operate on the series, similarly to how `(DataFrame|SeriesGroupby|DataFrameGroupBy).apply` operate on series. This is very fast and can do almost anything,
* `Series.map()` makes the passed-in callable operate on each series data elements individually. This is very flexible, but can be very slow, so should only be used if `Series.apply` can't do it.

so, this API change will help make Pandas `Series.(apply|map)` API  clearer without losing functionality and let their functionality  be explainable in a simple manner, which would be a win for Pandas.

The result from the above change will be that `Series.apply` will operate similar to how `DataFrame.apply` works already per column, similar to how `Series.map` operates similar to how `DataFrame.map` works per column. This will give better coherence between same-named methods on `DataFrames` and `Series`.

## Deprecation process

> To change the behavior to the current behavior will have to be deprecated. This can be done by adding a `by_row` parameter to `Series.apply`, which means, when `by_rows=False`, that `Series.apply` will not operate elementwise but Series-wise.

Reviewer 2: The question is what would be the default behavior if `by_row` is not specified?

Author: That will emit a deprecation warning. I've made text change to clarify this.
